JAVA-2254: Add support for computed properties

tolbertam opened the pull request:
For [JAVA-2254](https://datastax-oss.atlassian.net/browse/JAVA-2254)

Marking as wip as this needs to be rebased after #1239 is merged, and there's something I'm not completely satisfied with in `EntityHelperGetMethodGenerator` which I could use some feedback on.

---

Motivation:

It is useful to map values that are the result of some Cassandra-side
computation, typically function calls, to properties.

Modifications:

Introduce a `@Computed` annotation that can be applied to a field or
getter that allows specifying a formula to instruct Cassandra to
compute and return a value back that can be mapped back to this
property.

Introduce AliasedPropertyDefinition for enabling use of aliases in query
specification.  This is needed because the resulting value name for a
function call is not deterministic.

Add EntityDefinition.getComputedValues() for retrieving computed values
and added getAllValues(), which is similar to getAllColumns() but also
returns computed values.

Add PropertyDefinition.getCqlResultName() to differentiate between
request name and result name for alias definitions.

Update EntityHelperGetMethodGenerator to allow selectively retrieving
values when their cql name does not match their result name.

Update EntityHelperSelectStartMethodGenerator to include computed values
in generated select query.

Result:

`@Computed` annotation may be used to specify properties that are based
on result of some function call.  This is only usable by
`@Select`-annotated dao methods.

tolbertam committed:
JAVA-2254: Add support for computed properties

Motivation:

It is useful to map values that are the result of some Cassandra-side
computation, typically function calls, to properties.

Modifications:

Introduce a `@Computed` annotation that can be applied to a field or
getter that allows specifying a formula to instruct Cassandra to
compute and return a value back that can be mapped back to this
property.

Introduce AliasedPropertyDefinition for enabling use of aliases in query
specification.  This is needed because the resulting value name for a
function call is not deterministic.

Add EntityDefinition.getComputedValues() for retrieving computed values
and added getAllValues(), which is similar to getAllColumns() but also
returns computed values.

Add PropertyDefinition.getCqlResultName() to differentiate between
request name and result name for alias definitions.

Update EntityHelperGetMethodGenerator to allow selectively retrieving
values when their cql name does not match their result name.

Update EntityHelperSelectStartMethodGenerator to include computed values
in generated select query.

Result:

`@Computed` annotation may be used to specify properties that are based
on result of some function call.  This is only usable by
`@Select`-annotated dao methods.

tolbertam commented on manual/mapper/entities/README.md:
````diff
@@ -159,10 +159,43 @@ private int day;
+private int f;
+```
+
+Finally, note that [@Computed] fields are only used for [@Select]-annotated dao methods. Both 
````
Is this ok? Mapper 3.x was similar with `@Computed` not working with `@Accessor`, for which we opened [JAVA-832](https://datastax-oss.atlassian.net/browse/JAVA-832).   I'm not sure there is an easy way to resolve this without changing query contents.

One possible way to make that work is to alias the result to the name of the property.  We can then instruct others that when they make their queries to do the same, i.e.:

```
@Computed("ttl(x)")
private int xttl;
```

Would cause our select query to query this as `ttl(x) as xttl` instead of what i'm currently doing which is `ttl(x) as "[alias0]"`.

olim7t replied:
Yeah, I think it would make sense to use `getCqlName()` as the alias, and require that people do the  same in their custom queries.

tolbertam replied:
Awesome, definitely like this solution, its much more pleasant and also solves my other problem in `EntityHelperGetMethodGenerator` and also opens up computed to be used with the other query-based annotations :+1:

tolbertam replied:
Took a stab at that in 3767374 and I think it's a much more elegant solution overall.

I also made it work such that naming strategy / and @CqlName are regarded.

Updated tests to verify `@Query` and `@GetEntity` work assuming you make the right query, and updated the documentation to give instruction on how to do that.

tolbertam commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/EntityHelperGetMethodGenerator.java:
```diff
@@ -62,24 +62,34 @@ public EntityHelperGetMethodGenerator(
+      CodeBlock cqlResultName = property.getCqlResultName();
       String setterName = property.getSetterName();
       getBuilder.addCode("\n");
+      // if the result name is not the same as the name, it's likely this is a computed
```
Not really happy with this, namely because throwing exceptions can be costly, but with `GettableByName` there is no way to determine if a definition is present without throwing an exception (`firstIndexOf` will throw an `IllegalArgumentException`).   If this was a `Row`, i could use `getDefinitions`.

Alternatively if we based alias names on property names (as described [here](https://github.com/tolbertam)) we could just remove this logic and assume the value is present, and if not fail like we would do for any other column.

tolbertam replied:
this is no longer an issue as this special logic is not needed.

tolbertam commented:
Rebased on java2078, will proceed with changing how alias works

tolbertam committed:
Use property name for alias instead of generating names

tolbertam committed:
Doc cleanup

tolbertam commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/ComputedPropertyDefinition.java:
```diff
@@ -0,0 +1,74 @@
+            .orElse(cqlNameGenerator.buildCqlName(javaName));
+    // the queried name is "formula as alias"
+    this.cqlName =
+        CodeBlock.of("$S + $S +", formula, " as ").toBuilder().add(this.cqlResultName).build();
```
Is there a better way to do this?  The generated code looks like:

```java
    return selectFrom
        .column("id")
        .column("c_id")
        .column("v")
        .raw("writetime(v)" + " as " +"writetime")
        .raw("ttl(v)" + " as " +"myttl");
```

Which works, but would have been nice to do it as one clean string constant.

olim7t replied:
You can alias raw expressions:
```java
.raw("writetime(v)").as("writetime")
.raw("ttl(v)").as("myttl");
```

tolbertam replied:
Nice, that's definitely better.  It will require some reorienting of `ComputedPropertyDefinition` and how it is used in `EntityHelperSelectStartMethodGenerator`, but I should probably do that anyways.

tolbertam replied:
Was a little easier to do that than i originally anticipated (6b24969)

tolbertam committed:
writetime should be a long

tolbertam committed:
Improve query generation for computed properties

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/DefaultEntityFactory.java:
```diff
@@ -81,6 +81,8 @@ public EntityDefinition getDefinition(TypeElement classElement) {
     SortedMap<Integer, PropertyDefinition> clusteringColumns = new TreeMap<>();
     ImmutableList.Builder<PropertyDefinition> regularColumns = ImmutableList.builder();
+    ImmutableList.Builder<PropertyDefinition> computedValues = ImmutableList.builder();
+    int aliasCounter = 0;
```
Nit: unused

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/EntityDefinition.java:
```diff
@@ -54,4 +56,17 @@
+        .addAll(getRegularColumns())
+        .addAll(getComputedValues())
+        .build();
+  }
```
👍 the terminology makes sense.

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/ComputedPropertyDefinition.java:
```diff
@@ -0,0 +1,72 @@
+import com.squareup.javapoet.CodeBlock;
+import java.util.Optional;
+
+public class ComputedPropertyDefinition implements PropertyDefinition {
```
Do we really need to duplicate the whole class? It feels like adding an `Optional<String> formula` parameter to the original class would be pretty easy.

tolbertam replied:
:+1:, I went ahead and merged logic into `DefaultPropertyDefinition`

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/PropertyDefinition.java:
```diff
@@ -22,6 +22,10 @@
+  default CodeBlock getCqlResultName() {
+    return getCqlName();
+  }
+
```
I get which is which when I read this class, but in the rest of the code my head is exploding 😃 

First I think we need javadocs:
* `getCqlName`: a Java snippet that produces the corresponding expression in a SELECT statement, for example:
  * `"id"` in `selectFrom.column("id")` for a regular column
  * `"writetime(v)"` in `selectFrom.raw("writetime(v)")` for a computed value
* `getCqlResultName`: a Java snippet that produces the name of the property in a GettableByName or SettableByName, for example `"id"` in `row.get("id")`.

Then I would also rename them. Proposal:
* `getCqlName` => `getSelector`
* `getCqlResultName` => `getAlias`

tolbertam replied:
Sounds good, i was taking a path of least resistance initially.  Agree that renaming things makes things easier to understand.

tolbertam replied:
[More discussion here](https://github.com/datastax/java-driver/pull/1242#discussion_r288206815)

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/EntityHelperGetMethodGenerator.java:
```diff
@@ -62,9 +62,10 @@ public EntityHelperGetMethodGenerator(
-    for (PropertyDefinition property : entityDefinition.getAllColumns()) {
+    for (PropertyDefinition property : entityDefinition.getAllValues()) {
       PropertyType type = property.getType();
       CodeBlock cqlName = property.getCqlName();
```
`cqlName` shouldn't be referenced at all in this method. There are a couple of uses for UDTs and collections of UDTs below, but I think those should be changed to use `cqlResultName` too.

tolbertam replied:
After reorienting `PropertyDefinition`, `cqlName` is now appropriate again :+1:

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/EntityHelperGetMethodGenerator.java:
```diff
@@ -74,12 +75,12 @@ public EntityHelperGetMethodGenerator(
           //     returnValue.setLength(source.getInt("length"));
           getBuilder.addStatement(
-              "returnValue.$L(source.get$L($L))", setterName, primitiveAccessor, cqlName);
+              "returnValue.$L(source.get$L($L))", setterName, primitiveAccessor, cqlResultName);
```
There are other method generators where we should do that change, for example `EntityHelperInsertMethodGenerator`. In practice things work because they happen to deal exclusively with regular columns where `cqlName == cqlResultName`, but strictly speaking they should use `cqlResultName`.

If we do the `cqlName => selector` change that I'm suggesting in another comment, it will be more obvious and we can do a quick pass on the code.

tolbertam replied:
Correct me if I am mistaken, but I do not think this is relevant here now that we are renaming `getCqlResultName()` -> `getAlias()`.   Aliasing is only relevant in the context of select-based methods ([doc](https://github.com/apache/cassandra/blob/trunk/doc/cql3/CQL.textile#select-clause)), so it does not apply in the context insert/update/delete operations.

tolbertam replied:
After discussion we opted for `getSelector()` and `getCqlName()`, which makes it clear that `getCqlName` should be used here and most other places except when generating select query.

tomekl007 commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/ComputedIT.java:
```diff
@@ -0,0 +1,306 @@
+    // these should be set
+    assertThat(retrievedValue.getTtl()).isEqualTo(3600);
+    assertThat(retrievedValue.getWritetime()).isEqualTo(time);
+  }
```
I think we are missing test case that does `@Computed("ttl(v)")` and `@CqlName("myttl")` but cql name is not present in results from the query

tolbertam replied:
You mean just to validate that an exception is thrown if the user provided query does not match the expectation of what the result alias should be?

tomekl007 commented on integration-tests/src/test/java/com/datastax/oss/driver/mapper/ComputedIT.java:
```diff
@@ -0,0 +1,306 @@
+    ResultSet result =
+        session.execute(
+            SimpleStatement.newInstance(
+                "select id, c_id, v, writetime(v) as writetime, ttl(v) as myttl from "
```
Looking at this definition:
``` @CqlName("myttl")
    @Computed("ttl(v)")
    private int ttl;
```
Don't you think it is a little bit confusing? 
If the select query is returning `mytll` alias maybe it will be enough to do: `@Compued(ttl(myttl))`?

tolbertam replied:
i'm intentionally doing this to test that the `@CqlName` annotation is used over the property name:

```java

    @CqlName("myttl")
    @Computed("ttl(v)")
    private int ttl;
```

I'll update the comment to reflect that.

tolbertam committed:
update comment on test

olim7t reviewed (changes requested):
Looking really good, a few comments about naming.

adutra commented on manual/mapper/entities/README.md:
````diff
@@ -159,6 +159,67 @@ private int day;
+private int f;
+```
+
+[@Computed] fields are only used for queries, so they will not be considered for [@Update] or
````
To clarify I would add: `[@Computed] fields are only used for [@Select] queries, so they will not be considered for [@Update]...`

tolbertam replied:
It also works for `@GetEntity` and `@Query` as well (see sections below), so I think it makes sense as written, but can adjust it if you prefer.

adutra replied:
It's just that, at least to my ears, "queries" can refer to any form of CQL query, including inserts and updates. But OK to leave it as is if you think it's clear enough.

olim7t replied:
Maybe we could list them explicitly.

tolbertam replied:
updated to `select-based` queries so it's more clear.

adutra commented on manual/mapper/entities/README.md:
```diff
@@ -159,6 +159,67 @@ private int day;
+  @Computed("ttl(v)")
+  private int myTtl;
+
+  @Computed("writetime(v)")
```
Wouldn't `@Computed(expression = "writetime(v)", alias = "ts")` be more explicit?
If not: wouldn't `@Alias` be a better name than `@CqlName`?

olim7t replied:
> `@Computed(expression = "writetime(v)", alias = "ts")`

This introduces a special case though, isn't it better if computed values behave like any other property?

> wouldn't `@Alias` be a better name than `@CqlName`?

👍 to that. This is in line with the renaming I suggested [here](https://github.com/datastax/java-driver/pull/1242#discussion_r287579567).

tolbertam replied:
> Isn't it better if computed values behave like any other property?

That's what I was aiming for by using `@CqlName`.  `CqlName` is used to indicate the name of the cql column, but also now the aliased result name for computed values.  I think it fits really well.

If we change it to `@Alias`, the intent could be interpreted differently, effectively as a rename, but that will not generally be the case (only for `@Computed` properties).  I don't think there would be too much of a use case for adding something like `@Alias` to rename how to get at the resulting value beyond computed values.

olim7t replied:
> If we change it to `@Alias`, the intent could be interpreted differently

Mmm that's a good point, "alias" would imply that we use a different name than the column's, but for non-computed columns that is not the case.
So I change my vote to keep `CqlName`.

adutra replied:
I'm sorry but I find `CqlName` way too vague. If I follow you correctly this annotation is being used to:

1. Customize regular column names: why not introduce a special `@Column` annotation?
2. To define aliases for computed properties: why don't we revert then to my initial idea of `@Computed(expression = "writetime(v)", alias = "ts")`? It seems that only computed properties will ever need aliases, so that option sounds fair to me.

olim7t replied:
> why not introduce a special `@Column` annotation?

That sounds even more vague to me. Marking something as a "column" doesn't say anything about naming, and you might have other properties that are not annotated but will still map to a column.

> To define aliases for computed properties

The advantage of the current solution is that it works the same way whether the property is computed or not:

"What name should I use to get the property on my `Row`?"
* if `CqlName` is present, that
* otherwise, the name of the Java property, translated with the naming strategy

Why make it different for computed properties?

adutra replied:
You know what, I'm retiring myself from this endless debate and let you guys decide what is best. Good luck ;-)

adutra commented on mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/Computed.java:
```diff
@@ -0,0 +1,43 @@
+ * <p>Example:
+ *
+ * <pre>
+ * private int v;
```
This line seems unnecessary.

adutra commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/ComputedPropertyDefinition.java:
```diff
@@ -0,0 +1,72 @@
+
+  public ComputedPropertyDefinition(
+      String javaName,
+      Optional<String> customCqlName,
```
It is not a good practice to pass `Optional` as a method parameter, to avoid unnecessary wrapping/unwrapping (and also because `Optional` is not serializable). I know it's already done in a few other places, but wanted to pointed that out. A better approach is to pass the string annotated with `@Nullable`.

tolbertam replied:
I'll be factoring this code out into DefaultPropertyDefinition, will update the optional parameters in that constructor when I get to it.

olim7t replied:
> It is not a good practice to pass Optional as a method parameter

I don't 100% agree with that IDE warning. Performance is not paramount in the processor (at least not to the point where wrapping something in an optional will make a difference), and we don't store the optional in a field so serialization is not an issue (we don't serialize this object anyway for that matter).

Passing an optional here works nicely because the annotation parsing code that is used to fill that parameter already returns one.

adutra replied:
I agree with you, the only real bad thing is to store optionals as fields, but that is not the case here, so I'm fine if we keep them.

adutra commented:
`ComputedIT` seems to be failing against C* 2.1:

https://jenkins-drivers.build.dsinternal.org/view/java-driver/job/datastax.java-driver.PR.1242.default/8/testReport/

tolbertam committed:
Process feedback

* Update doc to make it more clear that `@Computed` is only applicable
  for select queries.
* Remove `ComputedPropertyDefinition` and merge logic into
  `DefaultPropertyDefinition`.
* Reorient `PropertyDefinition`. `getCqlName` will now map
  to the column name and result name used by `Gettable` and
  `SettableByName` and will be used as an alias in the case when a
  property is `@Computed`.  `getSelector` will be used when generating
  select queries only.
* Remove unused `aliasCounter` in `DefaultEntityFactory`.

tolbertam committed:
Merge branch 'java2078' into java2254

tolbertam committed:
Set nullSavingStrategy to SET_TO_NULL so test works with 2.1

This should be updated once there is a more global way to set the
strategy.

tolbertam committed:
Feedback: Add negative test for mismatched alias name

olim7t commented on mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/entity/EntityHelperSelectStartMethodGenerator.java:
```diff
@@ -56,8 +56,13 @@ public EntityHelperSelectStartMethodGenerator(
-    for (PropertyDefinition property : entityDefinition.getAllColumns()) {
-      selectStartBuilder.addCode("\n.column($L)", property.getCqlName());
+    for (PropertyDefinition property : entityDefinition.getAllValues()) {
+      if (property.getSelector().equals(property.getCqlName())) {
```
Nit: this does not clearly indicate that we distinguish computed values.

We could either introduce `PropertyDefinition.isComputed()`, or iterate `getAllColumns` and `getComputedValues` separately.

tolbertam replied:
That's actually how I previously had it, my reasoning for changing it was in possible anticipation of aliasing not necessarily being bound to computed properties, although that seems rather unlikely.

it probably is better to make it more explicit that this is why things are being aliased here, I'll change it back.

olim7t reviewed (approved):
LGTM

tolbertam committed:
Feedback: only alias computed properties
